src/connectionbenchmark/benchmarktaskdownload.cpp

```diff
@@ -31,43 +40,85 @@ void BenchmarkTaskDownload::handleState(BenchmarkTask::State state) {
   logger.debug() << "Handle state" << state;
 
   if (state == BenchmarkTask::StateActive) {
-    m_request = NetworkRequest::createForGetUrl(this, m_fileUrl);
-    connect(m_request, &NetworkRequest::requestUpdated, this,
+    // Start DNS resolution
+#if !defined(MVPN_DUMMY)
+    m_dnsLookup.setNameserver(QHostAddress(MULLVAD_DEFAULT_DNS));
+#endif
+    m_dnsLookup.lookup();
+  } else if (state == BenchmarkTask::StateInactive) {
+    for (NetworkRequest* request : m_requests) {
+      request->abort();
+    }
+    m_requests.clear();
+    m_dnsLookup.abort();
+  }
+}
+
+void BenchmarkTaskDownload::dnsLookupFinished() {
+  if (state() != BenchmarkTask::StateActive) {
+    logger.warning() << "DNS Lookup finished after task aborted";
```

Comment:
we still need to complete the task, otherwise, the task chain will be blocked.
I suspect you can treat this and the following if-stmt together.

Reply:
Okay, I think I have gotten the cancellation behaviour correct now. Calling `m_dnsLookup.abort()` should trigger the `finished()` signal with a status of `QDnsLookup::OperationCancelledError`. We've kept the `StateActive` check just in case we happen to catch a race condition where the user clicks cancel before a pending DNS lookup gets processed.